### PR TITLE
Debug Active Record Callbacks

### PR DESCRIPTION
**Pull request by andycroll**

### CODE REVIEW

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
+---
+title: "Fimd out what Callbacks are defined on an Active Record model in the console"
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *Find

Comment (nodunayo) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *callbacks

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+description: "Sometimes it's hars to know"
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *hard

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+layout: article
+category: ruby
+image:
+  base: "2022/find-list-debug-active-record-callbacks-in-the-console"
+  alt: "Three old payphones"
+  credit: "Pavan Trikutam"
+  source: "https://unsplash.com/photos/71CjSSB83Wo"
+---
+
+I was recently asked by my [First #RubyFriend mentee](https://firstrubyfriend.org) how to list the callbacks present on an Active Record model. I didn't know.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: I think 'mentee' shouldn't be included in the link! (A super minor point!)

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+The situation was a large legacy codebase with some large models constructed across multiple files, using various gems, custom callbacks and concerns. It was very hard to parse the callbacks by reading across multiple files.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *We/They were looking at a vast...with some large models defined (?) across...multiple files. They hooked into a variety of gems and featured custom callbacks and concerns.

Comment (nodunayo) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *parse just the callbacks across the multiple files.

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+So I found a mention of debugging callbacks in the [Rails API documentation](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html#module-ActiveRecord::Callbacks-label-Debugging+callbacks) and after a little digging built a simple loop you can use to find the callbacks defined on an Active Record model.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *I found a mention of...

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+...a loop in the Rails console.
+
+Open the console.
+
+```shell
+bin/rails console
+```
+
+Paste this code into the console to see the user-defined callbacks on a model.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *:

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    # remove autosaving callbacks from result
+    next if c.filter.to_s.include?("autosave")
+    next if c.filter.to_s.include?("_ensure_no_duplicate_errors")
+
+    result["#{c.kind}_#{c.name}"] += [c.filter]
+  end
+end
+```
+
+For the `User` object (which is powered by `devise`) in my application the results look like:
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
For the `User` object in my application (which is powered by [Devise](https://github.com/heartcombo/devise)) the results look like:
```

Does that look like your application is powered by Devise though? The parentheses just read a little awkwardly where they currently are...but maybe you just leave it.

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  "before_save"=>[:ensure_authentication_token],
+  "after_create"=>[:skip_reconfirmation_in_callback!],
+  "before_create"=>[:generate_confirmation_token],
+  "after_update"=>[:send_password_change_notification, :send_email_changed_notification],
+  "before_update"=>[:clear_reset_password_token :postpone_email_change_until_confirmation_and_regenerate_confirmation_token],
+  "after_commit"=>[:send_on_create_confirmation_instructions, :send_reconfirmation_instructions],
+}
+```
+
+Here you can see the callbacks inserted into the `User` model by our configuration of the `devise` gem. These aren't listed in the `app/models/user.rb` file.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: Hmmm..but earlier you say 'user-defined' callbacks...

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+You can also see a dynamic `Proc`-based callback inserted by our use of the `money-rails` gem.
+
+
+## How does this work?
+
+The `#__callbacks` method on a model is [defined as a Class method via Active Support](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/callbacks.rb#L68). If accessed directly (like `YourModel.__callbacks`) it contains all the details of the logic performed by an Active Record instance before or after changing its state.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *class

Comment (nodunayo) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *...accessed directly....like calling `Your Model...`, it returns...

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+The code above just pulls out the "headline" methdo calls
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *The code above only pulls out only the "headline" _method_ calls.

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+In the specific code above I’ve presumed the following:
+
+You aren't interested in seeing all the validations, like `validates_presence_of :attribute`, on a model these are stored in the callback chain as well. Hence `next if k = : validate`.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: Make it a bullet point list?

Comment (nodunayo) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: This is a run-on sentence? Needs more punctuation!

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+You aren't interested in the callbacks that automatically save the values of associated models where you `have_many :related_models`. Hence `next if c.filter.to_s.include?("autosave")`.
+
+
+## Why?
+
+This code can be useful when you’re new to a complex application or model or as a debugging tool.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: This sentence reads a little awkwardly!

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+I tend to limit my use of callbacks as much as possible due to the ease with which I‘ve confused myself before, but if you’re in an unfamilar applicaiton that uses them a lot, I hope this code helps your investigation.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: Maybe something like: helps you get up to speed/familiarise yourself it more quickly ?

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+
+## Why not?
+
+Please don’t use this code in production code, _only_ for debugging and understand a new codebase.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
Please don’t use this code in production. It's _only_ for debugging and helping you to understand a new codebase.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+I can't think of a reason to use it in application code off the top of my head, but I’m presuming this is mostly internal Rails API and is subject to change.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: Don't think this sentence is needed?

But if so...

"I can't think of a reason to use this in application code. Furthermore, it's mostly based on Rails's internal API and is subject to change."

Perhaps!

**Review by nodunayo — DISMISSED** [review]

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 1 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
title: "Find out what callbacks are defined on an Active Record model in the console"
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 2 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
description: "Sometimes it's hard to know what is defined where"
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 3 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
I was recently asked by my [First #RubyFriend](https://firstrubyfriend.org) mentee how to list the callbacks present on an Active Record model. I didn't know.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 4 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
They were looking at a vast legacy codebase with some large models defined across multiple files. The models used various gems, custom callbacks and concerns. It was very hard to parse what methods and behaviours the callbacks were causing.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 5 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
I found mention of [debugging callbacks in the Rails API documentation](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html#module-ActiveRecord::Callbacks-label-Debugging+callbacks) and after a little digging built a simple loop you can use to find the callbacks defined on an Active Record model.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 6 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
Paste this code into the console to see the user-defined callbacks on a model:
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 7 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
For the `User` object in my application, which uses [the Devise gem](https://github.com/heartcombo/devise) for authentication, the results look like:
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 9 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
The `#__callbacks` method on a model is [defined as a class method via Active Support](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/callbacks.rb#L68). If accessed directly, like calling `YourModel.__callbacks`, it return all the details of the logic performed by an Active Record instance before or after changing its state.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 11 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
You aren't interested in seeing all the validations: code like `validates_presence_of :attribute` creates a an entry in callback chain. We skip displaying these with `next if k == :validate`.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 12 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
When you’re new to a complex application this may be useful as a debugging tool.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 13 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
I tend to limit my use of callbacks as much as possible due to the ease with which I‘ve confused myself before, but if you’re in a code base that uses them a lot, this might get you up to speed with the existing behaviour.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 10 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
The code above only pulls out the "headline" method calls
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 8 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: We still defined them. They just come from devise.

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 15 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
I can't think of a reason to use this in application code. Furthermore, the `__` prefix in the method name indicates this is mostly based on Rails’s internal API and is subject to change.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+```
+
+Here you can see the callbacks inserted into the `User` model by our configuration of the `devise` gem. These aren't listed in the `app/models/user.rb` file.
+
+You can also see a dynamic `Proc`-based callback inserted by our use of the `money-rails` gem.
+
+
+## How does this work?
+
+The `#__callbacks` method on a model is [defined as a class method via Active Support](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/callbacks.rb#L68). If accessed directly, like calling `YourModel.__callbacks`, it return all the details of the logic performed by an Active Record instance before or after changing its state.
```
Comment (andycroll) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
The `#__callbacks` method on a model is [defined as a class method via Active Support](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/callbacks.rb#L68). If accessed directly, like calling `YourModel.__callbacks`, it returns all the details of the logic performed by an Active Record instance before or after changing its state.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+image:
+  base: "2022/find-list-debug-active-record-callbacks-in-the-console"
+  alt: "Three old payphones"
+  credit: "Pavan Trikutam"
+  source: "https://unsplash.com/photos/71CjSSB83Wo"
+---
+
+I was recently asked by my [First #RubyFriend](https://firstrubyfriend.org) mentee how to list the callbacks present on an Active Record model. I didn't know.
+
+They were looking at a vast legacy codebase with some large models defined across multiple files. The models used various gems, custom callbacks and concerns. It was very hard to parse what methods and behaviours the callbacks were causing.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: Oxford comma?? 🥺

Comment (nodunayo) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: "...what methods" sounds a little awkward alongside "causing". Maybe just stick to "behaviours"?

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+I found mention of [debugging callbacks in the Rails API documentation](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html#module-ActiveRecord::Callbacks-label-Debugging+callbacks) and after a little digging built a simple loop you can use to find the callbacks defined on an Active Record model.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
I found mention of [debugging callbacks in the Rails API documentation](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html#module-ActiveRecord::Callbacks-label-Debugging+callbacks) and, after a little digging, built a simple loop you can use to find the callbacks defined on an Active Record model.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+    # remove autosaving callbacks from result
+    next if c.filter.to_s.include?("autosave")
+    next if c.filter.to_s.include?("_ensure_no_duplicate_errors")
+
+    result["#{c.kind}_#{c.name}"] += [c.filter]
+  end
+end
+```
+
+For the `User` object in my application, which uses [the Devise gem](https://github.com/heartcombo/devise) for authentication, the results look like:
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: Yess!

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 8 above)
Comment (nodunayo) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: True!

Comment (nodunayo) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *based on our configuration...?

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+Here you can see the callbacks inserted into the `User` model by our configuration of the `devise` gem. These aren't listed in the `app/models/user.rb` file.
+
+You can also see a dynamic `Proc`-based callback inserted by our use of the `money-rails` gem.
+
+
+## How does this work?
+
+The `#__callbacks` method on a model is [defined as a class method via Active Support](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/callbacks.rb#L68). If accessed directly, like calling `YourModel.__callbacks`, it returns all the details of the logic performed by an Active Record instance before or after changing its state.
+
+The code above only pulls out the "headline" method calls
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *.

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+
+In the specific code above I’ve presumed the following:
+
+You aren't interested in seeing all the validations: code like `validates_presence_of :attribute` creates a an entry in callback chain. We skip displaying these with `next if k == :validate`.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
You aren't interested in seeing all of the validations: code like `validates_presence_of :attribute` creates an entry in the callback chain. We skip displaying these with `next if k == :validate`.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+
+You aren't interested in the callbacks that automatically save the values of associated models where you `have_many :related_models`. Hence `next if c.filter.to_s.include?("autosave")`.
+
+
+## Why?
+
+When you’re new to a complex application this may be useful as a debugging tool.
+
+I tend to limit my use of callbacks as much as possible due to the ease with which I‘ve confused myself before, but if you’re in a code base that uses them a lot, this might get you up to speed with the existing behaviour.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *codebase?

Comment (nodunayo) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *..existing behaviour more quickly?

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+
+
+## Why not?
+
+Please don’t use this code in production. It's _only_ for debugging and helping you to understand a new codebase.
+
+I can't think of a reason to use this in application code. Furthermore, the `__` prefix in the method name indicates this is mostly based on Rails’s internal API and is subject to change.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md:89]: Much better!

**Review by nodunayo — DISMISSED** [review]

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 17 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
They were looking at a vast legacy codebase, with some large models defined across multiple files. The models used various gems, custom callbacks, and concerns. It was very hard to parse what behaviours the callbacks were causing.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 20 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
The code above only pulls out the "headline" method calls.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 22 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
I tend to limit my use of callbacks as much as possible due to the ease with which I‘ve confused myself before, but if you’re in a codebase that uses them a lot, this might get you up to speed with the existing behaviour more quickly.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 8 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
Here you can see the callbacks inserted into the `User` model based on our configuration of the `devise` gem. These aren't listed in the `app/models/user.rb` file.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+    # remove autosaving callbacks from result
+    next if c.filter.to_s.include?("autosave")
+    next if c.filter.to_s.include?("_ensure_no_duplicate_errors")
+
+    result["#{c.kind}_#{c.name}"] += [c.filter]
+  end
+end
+```
+
+For the `User` object in my application, which uses [the Devise gem](https://github.com/heartcombo/devise) for authentication, the results look like:
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *like this

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+```
+
+Here you can see the callbacks inserted into the `User` model based on our configuration of the `devise` gem. These aren't listed in the `app/models/user.rb` file.
+
+You can also see a dynamic `Proc`-based callback inserted by our use of the `money-rails` gem.
+
+
+## How does this work?
+
+The `#__callbacks` method on a model is [defined as a class method via Active Support](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/callbacks.rb#L68). If accessed directly, like calling `YourModel.__callbacks`, it returns all the details of the logic performed by an Active Record instance before or after changing its state.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: Not sure about the second sentence here, namely not sure it's clear what 'all the details of the logic performed...' means.

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+
+The code above only pulls out the "headline" method calls.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: * and I've presumed the following:

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+
+In the specific code above I’ve presumed the following:
+
+You aren't interested in seeing all of the validations: code like `validates_presence_of :attribute` creates an entry in the callback chain. We skip displaying these with `next if k == :validate`.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: *validations. Code like...

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+
+You aren't interested in the callbacks that automatically save the values of associated models where you `have_many :related_models`. Hence `next if c.filter.to_s.include?("autosave")`.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md:75]: Think this could be a little clearer!

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+
+
+## Why?
+
+When you’re new to a complex application this may be useful as a debugging tool.
+
+I tend to limit my use of callbacks as much as possible due to the ease with which I‘ve confused myself before, but if you’re in a codebase that uses them a lot, this might get you up to speed with the existing behaviour more quickly.
```
Comment (nodunayo) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md:82]: I wish this paragraph was a little less "I"-centric, and more focused on the potential pain for the reader.

**Review by nodunayo — CHANGES_REQUESTED** [review]

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 26 above)
Comment (juliachn) [inline _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: looking good 👍🏽  this is the only part I'm a bit confused about - any chance to include a short line what is "headline" method?

Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
The code above only pulls out the main method calls.
```

Comment (juliachn) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]: thanks, much better! At least I could find a definition on google

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 24 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
For the `User` object in my application, which uses [the Devise gem](https://github.com/heartcombo/devise) for authentication, the results look like this:
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 25 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
The `#__callbacks` method on a model is [defined as a class method via Active Support](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/callbacks.rb#L68). If accessed directly, like calling `YourModel.__callbacks`, it returns a complex hash structure containing the logic performed by the Active Record instance before or after changing its state.
```

File: _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md (hunk 27 above)
Comment (andycroll) [reply _posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md]:
```suggestion
You aren't interested in seeing all of the validations. Code like `validates_presence_of :attribute` creates an entry in the callback chain. We skip displaying these with `next if k == :validate`.
```
